src/audio/base_fw.c

```diff
@@ -241,23 +243,46 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
 	return 0;
 }
 
+static log_timestamp_t basefw_get_timestamp(void)
+{
+	return sof_cycle_get_64() + cycle_delta;
+}
+
 static uint32_t basefw_set_system_time(uint32_t param_id,
 				       bool first_block,
 				       bool last_block,
 				       uint32_t data_offset,
 				       const char *data)
 {
-	/* TODO: configurate time to logging subsystem */
+	uint64_t dsp_time;
+	uint64_t dsp_cycle;
+	uint64_t host_time;
+	uint64_t host_cycle;
+
 	if (!(first_block && last_block))
 		return IPC4_INVALID_REQUEST;
 
 	global_system_time_info.host_time.val_l = ((const struct ipc4_system_time *)data)->val_l;
 	global_system_time_info.host_time.val_u = ((const struct ipc4_system_time *)data)->val_u;
 
-	uint64_t current_dsp_time = k_cyc_to_us_floor64(sof_cycle_get_64());
+	dsp_cycle = sof_cycle_get_64();
+	dsp_time = k_cyc_to_us_floor64(dsp_cycle);
+
+	global_system_time_info.dsp_time.val_l = (uint32_t)(dsp_time);
+	global_system_time_info.dsp_time.val_u = (uint32_t)(dsp_time >> 32);
+
+	/* use default timestamp if 64bit is not enabled since 64bit is necessary for host time */
+	if (!IS_ENABLED(CONFIG_LOG_TIMESTAMP_64BIT)) {
+		LOG_WRN("64bits timestamp is disabled, so use default timestamp");
+		return IPC4_SUCCESS;
+	}
 
-	global_system_time_info.dsp_time.val_l = (uint32_t)(current_dsp_time);
-	global_system_time_info.dsp_time.val_u = (uint32_t)(current_dsp_time >> 32);
+	host_time = global_system_time_info.host_time.val_l |
+			((uint64_t)global_system_time_info.host_time.val_u << 32);
+	host_cycle = k_us_to_cyc_ceil64(host_time);
+	cycle_delta = host_cycle - dsp_cycle;
```

Comment:
It took me a moment to figure out that host_time and host_cycle are local and cycle_delta is global. A minor thing, but given so many similarly named variables, naming cycle_delta "global_cycle_delta" would increase readability.

Reply:
thanks, good idea!